components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/ChallengeQuestionManager.java

```diff
@@ -558,30 +586,42 @@ public void setChallengesOfUser(User user, UserChallengeAnswer[] userChallengeAn
             // check whether the answered questions exist in the tenant domain
             checkChallengeQuestionExists(userChallengeAnswers, tenantDomain);
 
+            // Get the existing challenge questions and answers for the user.
+            Map<String, String> existingQuestionAndAnswers = retrieveAnsweredChallenges(user, userChallengeAnswers);
+
+            triggerChallengeAnswersValidation(user, userChallengeAnswers,
+                    existingQuestionAndAnswers, IdentityEventConstants.Event.PRE_SET_CHALLENGE_QUESTION_ANSWERS);
+
             List<String> challengesUris = new ArrayList<String>();
             String challengesUrisValue = "";
             String separator = getChallengeSeparator();
 
+            Map<String, String> challengeQuestionToUpdate = new HashMap<>();
+
             if (!ArrayUtils.isEmpty(userChallengeAnswers)) {
                 for (UserChallengeAnswer userChallengeAnswer : userChallengeAnswers) {
-                    if (userChallengeAnswer.getQuestion().getQuestionSetId() != null && userChallengeAnswer.getQuestion().getQuestion() !=
-                            null && userChallengeAnswer.getAnswer() != null) {
-                        String oldValue = Utils.
-                                    getClaimFromUserStoreManager(user, userChallengeAnswer.getQuestion().getQuestionSetId().trim());
+
+                    if (StringUtils.isNotBlank(userChallengeAnswer.getQuestion().getQuestionSetId()) &&
```

Comment:
Better to extract `userChallengeAnswer.getQuestion()` to a variable, since it is called again in multiple places.

Reply:
Fixed in https://github.com/wso2-extensions/identity-governance/pull/426/commits/50aebef76b14e869930e8c7582307d9d548c109e